### ISSUE TITLE
Problem when using with Angular

### ISSUE THREAD

**Issue by mustafaekim** [edcarroll/ta-json#18]
Hi,

I was using the library on nodejs and everything was working as expected. Now I try to do the same thing on angular (with webpack) and somehow the library just breaks everything. I cannot see the real error but wherever I load the JSON (from "ta-json") Angular fails to load the whole file.

> Uncaught Error: Unexpected value 'AppComponent' declared by the module 'AppModule'

I am getting the error above but I tried and tested many times that it only occurs if I use a simple JSON.deserialize or any other method it provides.

Any idea?

**Comment by mustafaekim** [comment]
It complains that JSON is not exported..

Error info:

```ts
'JSON' is not exported by node_modules\ta-json\index.js
https://github.com/rollup/rollup/wiki/Troubleshooting#name-is-not-exported-by-module
src\aot\weby\peoplive\spa\src\weby\account\sign-up-component.js (15:9)
13: import { AccountService } from "../../service/account.service";
14: import { Pattern } from "../../common/pattern";
15: import { JSON } from "ta-json";
```

**Comment by edcarroll** [comment]
This is most likely due to the fact that `ta-json` uses the CommonJS module format (so as to provide compatibility with NodeJS) - and so rollup complains a bit.

I think what you need to do is install the rollup CommonJS module, then [read this](https://github.com/rollup/rollup/wiki/Troubleshooting#name-is-not-exported-by-module), and add the parts of `ta-json` that you use, like `JSON`, `@JsonObject` etc to `namedExports`.

Let me know if that doesn't work

**Comment by mustafaekim** [comment]
That's the error I could get when I tried to compile with rollup/aot. I tried the recommendations on the link, could'not make it work

During development I use webpack and it silently ignores the library either.

**Comment by mustafaekim** [comment]
Now, I made it work with webpack. And on rollup, I figured out to skip "ta-json" export error but now I am getting Reflect.js export error. The error below is only thrown if I use ta-json library. I think I am lost here :)

'default' is not exported by 'node_modules\reflect-metadata\Reflect.js'

**Comment by edcarroll** [comment]
Hi there, sorry completely missed this. Were you able to resolve it?

**Comment by mustafaekim** [comment]
unfortunately no. I moved back to my old pattern (handling pure json) because it added a lot of errors like those, which I could not afford enough time to overcome

**Comment by mustafaekim** [comment]
actually since the beginning of my Typescript career (~1 year) I always though that I should transfer every json into a proper class instance, i still think that is the right way, however I observe that community prefer to build very advanced architecture build on transfering pure json objects/strings all around. I wanted to experience this bizarre pattern which sounds very disturbing to me, but I will try any way. Maybe because community also is experiencing several issues with json to class instance conversion.

**Comment by mustafaekim** [comment]
If I prefer to use typed class instances instead of basic json objects, here are the places where I should implement it throughout the project:

Whenever the client (angular) is sending a request, I convert my typed class instance into json.
The server is converting back the json to the typed class instance. If there is any database operation, I convert the typed class instance to json and persist it. If I need to get some data from database, I get it as json, then converts to typed class instance, do some stuff and convert back to json to send to my angular client.

lots of conversion. what do you think?

client (angular) <-> CLASS to JSON <-> SERVER (Node or Spring) <-> CLASS TO JSON <-> OBJECT DB (MongoDB)

**Comment by edcarroll** [comment]
Again, apologies for my slow replies, things have been pretty crazy.

That seems reasonable to me, as long as you aren't doing hundreds of these every second! I do basically the same thing (as it is a lot more powerful than just pure JSON passing).

On the topic of this issue, I've just released `2.3.0` which adds support for ES6 modules syntax - so you should be able to easily use the library with rollup now.

Could you give it a go and let me know how you get on?

Thanks!

**Comment by edcarroll** [comment]
Closing due to inactivity.
